src/qengine/state/opencl.cpp

```diff
@@ -93,68 +103,70 @@ void QEngineOCL::DispatchCall(
-    cmplx[4] = complex((bitCount == 1) ? (1.0 / runningNorm) : 1.0, 0.0);
+    cmplx[4] =
+        complex((doNormalize && (bitCount == 1) && (runningNorm > min_norm)) ? (1.0 / sqrt(runningNorm)) : 1.0, 0.0);
     bitCapInt bciArgs[BCI_ARG_LEN] = { bitCount, maxQPower, offset1, offset2, 0, 0, 0, 0, 0, 0 };
     for (int i = 0; i < bitCount; i++) {
         bciArgs[4 + i] = qPowersSorted[i];
     }
 
     /* Slightly different call parameters than the rest of the calls. */
-    queue.enqueueUnmapMemObject(stateBuffer, stateVec);
-    queue.enqueueWriteBuffer(cmplxBuffer, CL_FALSE, 0, sizeof(complex) * CMPLX_NORM_LEN, cmplx);
-    queue.enqueueWriteBuffer(ulongBuffer, CL_FALSE, 0, sizeof(bitCapInt) * BCI_ARG_LEN, bciArgs);
+    queue->enqueueUnmapMemObject(*stateBuffer, stateVec);
+    queue->enqueueWriteBuffer(cmplxBuffer, CL_FALSE, 0, sizeof(complex) * CMPLX_NORM_LEN, cmplx);
+    queue->enqueueWriteBuffer(ulongBuffer, CL_FALSE, 0, sizeof(bitCapInt) * BCI_ARG_LEN, bciArgs);
     if (doCalcNorm) {
         nrmParts = new real1[CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE]();
-        queue.enqueueWriteBuffer(
+        queue->enqueueWriteBuffer(
             nrmBuffer, CL_FALSE, 0, sizeof(real1) * CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE, nrmParts);
     }
-    queue.finish();
+    queue->finish();
 
     cl::Kernel apply2x2;
     if (doCalcNorm) {
-        apply2x2 = *(clObj->GetApply2x2NormPtr());
+        apply2x2 = *(clObj->GetApply2x2NormPtr(queue));
     } else {
-        apply2x2 = *(clObj->GetApply2x2Ptr());
+        apply2x2 = *(clObj->GetApply2x2Ptr(queue));
     }
-    apply2x2.setArg(0, stateBuffer);
+    apply2x2.setArg(0, *stateBuffer);
     apply2x2.setArg(1, cmplxBuffer);
     apply2x2.setArg(2, ulongBuffer);
     if (doCalcNorm) {
         apply2x2.setArg(3, nrmBuffer);
     }
-    queue.enqueueNDRangeKernel(apply2x2, cl::NullRange, // kernel, offset
+    queue->enqueueNDRangeKernel(apply2x2, cl::NullRange, // kernel, offset
         cl::NDRange(CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE), // global number of work items
         cl::NDRange(1)); // local number (per group)
 
-    queue.enqueueMapBuffer(stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(complex) * maxQPower);
-    if (doCalcNorm) {
-        queue.enqueueReadBuffer(
+    queue->enqueueMapBuffer(*stateBuffer, CL_TRUE, CL_MAP_READ | CL_MAP_WRITE, 0, sizeof(complex) * maxQPower);
+    if (doNormalize && doCalcNorm) {
+        queue->enqueueReadBuffer(
             nrmBuffer, CL_TRUE, 0, sizeof(real1) * CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE, nrmParts);
         runningNorm = 0.0;
         for (unsigned long int i = 0; i < CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE; i++) {
             runningNorm += nrmParts[i];
         }
         delete[] nrmParts;
-        runningNorm = sqrt(runningNorm);
-    } else {
-        runningNorm = 1.0;
     }
 }
 
 bitLenInt QEngineOCL::Cohere(QEngineOCLPtr toCopy)
 {
+    // LockGuard locked_call(deviceMutexPtr);
```

Comment:
This needs clean-up, because I was hacking it for debugging, but one of many QEnigineOCLs running on the same physical OCL device each need to lock temporary private ownership of their OCL device. If they don't have private ownership of the device, physical buffers are written to and queued as kernel arguments in the middle of different kernel initializations and calls, and it will produce garbage.

Reply:
This lock guard was temporarily disabled for debugging. I've re-enabled it. Yes, the mutex always needs to protect the queue.